Add 'and' and 'or' operators

JEbertPrime opened the pull request:
-And and Or logical operators added. No tests yet, but as of now they adhere to the behavior described by the wiki.
-Whitespace support added. In parser, use '_' to denote a space. This allows for the syntax of logical operators to be used accurately ('expression _ operator _ expression')
```JS
//Parser syntax

...head: Expression 
tail: (_ LogicalOperator _ Expression)...

//TI-BASIC Code
1<2 and 2>1
//Evaluates to 1 (true)
```

JEbertPrime committed:
Added 'and' and 'or' operators

-And and Or logical operators added. No tests yet, but as of now they adhere to the behavior described by the wiki.
-Whitespace support added. In parser, use '_' to denote a space. This allows for the syntax of logical operators to be used accurately ('expression _ operator _ expression')

JEbertPrime committed:
Removed console.log statement

davidtorosyan commented on src/parse/tibasic.pegjs:
```diff
@@ -429,4 +441,7 @@ Statement
\ No newline at end of file
+  / ValueStatement
+  
+  _ "whitespace"
+  = [ \t\n\r]*
```
I don't think this is the right approach. Whitespace is TI-BASIC is meaningful, so while these are valid:
```
1+1
1 and 1
```

These are syntax errors:
```
1+ 1
1 and  1
```

This means we can't collapse whitespace into a single label this way, and we certainly can't include newlines in there.

I think for `and` and `or` the spaces can be built into the token itself, just like `If `.

JEbertPrime replied:
Agreed - my thinking was that an explicit whitespace declaration might make error handling easier, but I think your approach is as good and removes lots of repetition.

davidtorosyan commented on src/parse/pegutil.js:
```diff
@@ -3,8 +3,20 @@
 
 import * as types from '../common/types'
 
+export function buildLogicalExpression (head, tail) {
```
I'm guessing this was needed because of the extra whitespace? I think if you include those in the token (see other comment) you can use `buildBinaryExpression` instead.

JEbertPrime replied:
See above comment resolving this

davidtorosyan commented on src/evaluate/expression.js:
```diff
@@ -225,6 +225,8 @@ function applyBinaryOperation (operator, x, y) {
     case '<=': return x <= y ? 1 : 0
     case '<': return x < y ? 1 : 0
+    case 'and': return x && y ? 1 : 0
+    case 'or': return x || y ? 1 : 0
```
`xor` might be a good candidate to include in this PR, since it follows the same pattern as these two (unlike `not(`). Up to you!

JEbertPrime replied:
xor is added!

davidtorosyan reviewed (changes requested):
Thanks for the PR!

Overall looks good, other than the comments I left. I think you're right about the precedence, and this should fail correctly on strings.

Would you mind adding some test cases for these as well? I generally look for some happy-path cases and some potential syntax and data type errors. I don't have test automation set up yet unfortunately, but if you update the test file I'll take your word for it that all the tests are still passing.

JEbertPrime committed:
Addresses Comments

-Whitespace denoted by `_` is removed. I agree about your approach; my thinking was that by making whitespace explicitly included in the parsing, we might catch errors more easily. But the cases are so limited that `' and '` is effective.
-My new function to parse whitespace is removed, and the `buildBinaryExpression` function is used.
-`xor` is added
-Test cases for all new operators have been added and are passing!

JEbertPrime commented:
I've addressed all your comments in the latest commit, and added test cases. This PR should be good to go!

davidtorosyan commented on src/evaluate/expression.js:
```diff
@@ -225,6 +225,9 @@ function applyBinaryOperation (operator, x, y) {
     case '<': return x < y ? 1 : 0
+    case ' and ': return x && y ? 1 : 0
+    case ' or ': return x || y ? 1 : 0
+    case ' xor ': return (!x && y) || (x && !y) ? 1:0
```
Note to self: set up a linter. No need to block on this though.

davidtorosyan reviewed (approved):
Woo! Looks good.
